File: ros2web/api/models/interface.py
from dataclasses import dataclass, field
from typing import Optional, ClassVar, Union
from typing import List, Any
from .data import Data


@dataclass
class Field:
    type: str
    name: str
    nested: Optional[List['Field']] = field(default=None)


@dataclass
class Msg(Data):
    type: ClassVar[str] = "msg"
    name: str
    package_name: str
    fields: Optional[List[Field]] = field(default=None)

    def __post_init__(self):
        self.id = f'{self.package_name}/{self.type}/{self.name}'


@dataclass
class Srv(Data):
    type: ClassVar[str] = "srv"
    name: str
    package_name: str
    request: Optional[List[Field]] = field(default=None)
    response: Optional[List[Field]] = field(default=None)

    def __post_init__(self):
        self.id = f'{self.package_name}/{self.type}/{self.name}'


@dataclass
class Act(Data):
    type: ClassVar[str] = "action"
    name: str
    package_name: str
    goal: Optional[List[Field]] = field(default=None)
    result: Optional[List[Field]] = field(default=None)
    feedback: Optional[List[Field]] = field(default=None)

    def __post_init__(self):
        self.id = f'{self.package_name}/{self.type}/{self.name}'

File: ui/src/models/interface.ts
import { Data } from ".";

export type Field = {
  type: string;
  name: string;
  nested?: Field[];
};

export type Msg = {
  name: string;
  packageName: string;
  fields?: Field[];
} & Data;

export type Srv = {
  name: string;
  packageName: string;
  request?: Field[];
  response?: Field[];
} & Data;

export type Act = {
  name: string;
  packageName: string;
  goal?: Field[];
  result?: Field[];
  feedback?: Field[];
} & Data;
